Issue 44682: Run PX data validation as a pipeline job

vagisha opened the pull request:
#### Rationale
We would like to run the data validation step of submitting a dataset to Panorama Public as a pipeline job.  The details are described in this doc: https://docs.google.com/document/d/1zDJY5f8rbb1Nq95__JxFStVmPqtcufFkVV3FY_wNNw0/edit

#### Related Pull Requests
* https://github.com/LabKey/targetedms/pull/497

#### Changes
- Run data validation for ProteomeXchange as a pipeline job
- Results from running the job are stored in new tables added to the panoramapublic schema
- PxValidationStatusApiAction returns the validation results as JSON which is displayed in pxValidationStatus.jsp
- CopyExperimentAction, PX actions (validate / submit / update PX XML), and PX XML and HTML writers have been updated to use the new data validation classes
- A spectral library is marked as incomplete if:
  - the SpectrumSourceFiles table of the .blib does not have the idFileName column, or there are now rows in the SpectrumSourceFiles table
  - the .blib library was built with an assay library (.csv files as source)
  - library is unsupported (only .blib and .elib are supported)
  - the library is referenced in the Skyline XML but not included in the .sky.zip
- For .blib built with MaxQuant search results, look for mqpar.xml etc. in addition to what is listed in the SpectrumSourceFiles.idFileName column
- If submitted data is in sub-folders we look for spectral library source files in the parent experiment folder as well as the sub-folders that have a Skyline document with the library. The files have to be uploaded to only one folder
- Sample files in a Skyline document that have the same file name but were imported from different paths will be marked as ambiguous, and the dataset will not be considered a valid PX dataset.  We may support this later by requiring that such files be uploaded into subfolders named for the replicate so that we can can link to the correct sample file.
- Added some fixes to matching modifications to Unimod in the case that a modification in the Skyline document does not have a Unimod ID. Updated test in ExperimentModificationGetter with modifications found on PanoramaWeb without a Unimod Id.
-  ExperimentUserForeignKey extends UserIdForeignKey so that the Submitter and LabHead columns can include all users when the experiment list is viewed in the Panorama Public project.
- Updated existing tests and added a PanoramaPublicValidationTest
- Moved test sample data to TargetedMS/panoramapublic subfolder

vagisha committed:
Issue 44682: Run PX data validation as a pipeline job
- Run data validation for ProteomeXchange as a pipeline job
- Results from running the job are stored in new tables added to the panoramapublic schema
- PxValidationStatusApiAction returns the validation results as JSON which is displayed in pxValidationStatus.jsp
- Updated CopyExperimentAction, PX actions (validate / submit / update PX XML), and PX XML and HTML writers to use the new data validation classes
- Spectral libraries are marked as incomplete if:
-- libraries that are not included in the sky.zip but are referenced in the Skyline XML
-- old libraries without an idFileName column in the SpectrumSourceFiles
-- libraries built with assay libraries (.csv files as source)
-- unsupported libraries: anything other than .blib or .elib
- For .blib built with MaxQuant search results, look for mqpar.xml etc. in addition to what is listed in the SpectrumSourceFiles table
- If submitted data is in sub-folders we look for spectral library source files in the parent experiment folder as well as the sub-folders that have a Skyline document with the library
- Added some fixes to matching modifications to Unimod in case the modification in the Skyline document does not have a Unimod ID. Updated test in ExperimentModificationGetter with several modifications found on PanoramaWeb without a Unimod Id.
- Sample files in a Skyline document that have the same file name but were imported from different paths will be marked as ambiguous
- Updated existing tests and added a PanoramaPublicValidationTest
- Moved test sample data to TargetedMS/panoramapublic subfolder
- In SpecLibReader.readLibSourceFiles() return empty list instead of throwing exception if the library file size is 0-bytes.
- ExperimentUserForeignKey extends UserIdForeignKey so that the Submitter and LabHead columns can include all users when the experiment list is viewed in the Panorama Public project.

vagisha committed:
Fix for failing PanoramaPublicValidationTest.

vagisha committed:
Expand rows as needed in DataValidationPage. Display '-' instead of '0' in the size column for libraries missing in the sky.zip.

labkey-jeckels commented on panoramapublic/resources/schemas/dbscripts/postgresql/panoramapublic-22.000-22.001.sql:
```diff
@@ -0,0 +1,141 @@
+    Name                       VARCHAR(300) NOT NULL, -- Name of the Skyline file
+
+    CONSTRAINT PK_SkylineDocValidation PRIMARY KEY (Id),
+    CONSTRAINT FK_SkylineDocValidation_DataValidation FOREIGN KEY (ValidationId) REFERENCES panoramapublic.DataValidation(Id)
```
Should probably have a FK on container too

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/PanoramaPublicController.java:
```diff
@@ -2541,46 +2874,447 @@ public void setDataLicense(String dataLicense)
+            return !_getPxid;
+        }
+
+//        public boolean isSkipPxCheck()
```
Can these commented lines be deleted?

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/PanoramaPublicSchema.java:
```diff
@@ -154,6 +173,118 @@ public TableInfo createTable(String name, ContainerFilter cf)
+
+        if (TABLE_SKYLINE_DOC_VALIDATION.equalsIgnoreCase(name))
+        {
+            var table = new PanoramaPublicTable(PanoramaPublicManager.getTableInfoSkylineDocValidation(), this, cf, ContainerJoin.DataValidatoinJoin);
```
Typo in ContainerJoin.DataValidatoinJoin

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/PanoramaPublicSchema.java:
```diff
@@ -204,11 +335,54 @@ public SQLFragment getFromSQL(String alias)
         }
-        return super.createView(context, settings, errors);
+
+        if (TABLE_SKYLINE_DOC_VALIDATION.equalsIgnoreCase(settings.getQueryName())
```
These tables could also override supportsContainerFilter() which might hide the controls in more places.

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/proteomexchange/ExperimentModificationGetter.java:
```diff
@@ -363,85 +635,154 @@ public void testStructuralMods() throws IOException
 
-                notFound = notFound + (pxMod.hasUnimodId() ? 0 : 1);
-                total++;
+//                if (pxMod.hasUnimodId())
```
Can this be trimmed?

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/publish/pxValidationStatus.jsp:
```diff
@@ -0,0 +1,852 @@
+        }
+
+        function getStatusDetails(statusId, json) {
+            var html =  statusId === 3 ? getStatusValidHtml(json)
```
Is it worth pulling these values into constants, even if they're just within this JSP?

labkey-jeckels reviewed (approved):
A few minor suggestions

vagisha committed:
- Code review changes
-- PanoramaPublicTable overrides supportsContainerFilter(); returns false if noContainerFilter is set.
-- Added constants for statusId in pxValidationStatus.jsp
- Bug fixes
-- Populate form with value of incompletePxSubmission when doing a resubmit or update
-- Use the correct container with AuditLogService.get().getAuditEvents()
